src/vue/modules/document-upload-form/index.vue

```diff
@@ -0,0 +1,200 @@
+import { initConfig } from './_config'
+
+import { Bus } from '@/js/helpers/event-bus'
+import { ErrorHandler } from '@/js/helpers/error-handler'
+
+import { FileUtil } from '@/js/utils/file.util'
+import { CryptoUtil } from './utils/crypto.util'
+
+import { DocumentUploader } from './helpers/document-uploader'
+
+import { mapActions, mapMutations } from 'vuex'
+import { types } from './store/types'
+
+const EVENTS = {
+  submit: 'submit',
+}
+
+export default {
+  name: 'document-upload-form-module',
+  mixins: [FormMixin],
+  props: {
+    /**
+     * @property config - the config for component to use
+     * @property config.horizonURL - the url of horizon server (without version)
+     * @property config.storageURL - the url of file storage server
+     */
+    config: {
+      type: Object,
+      required: true,
+    },
+    wallet: {
+      type: Wallet,
+      required: true,
+    },
+  },
+
+  data: _ => ({
+    form: {
+      document: null,
+      description: '',
+    },
+    isDocumentUploading: false,
+    DOCUMENT_TYPES,
+  }),
+  validations: {
+    form: {
+      document: { documentContainer },
+      description: { required },
+    },
+  },
+
+  created () {
+    initApi(this.wallet, this.config)
+    initConfig(this.config)
+    this.setAccountId(this.wallet.accountId)
+  },
+
+  methods: {
+    ...mapMutations('upload-form', {
+      setAccountId: types.SET_ACCOUNT_ID,
+      setDocumentAccountId: types.SET_DOCUMENT_ACCOUNT_ID,
+    }),
+
+    ...mapActions('upload-form', {
+      createAccount: types.CREATE_ACCOUNT,
+      createChangeRoleRequest: types.CREATE_CHANGE_ROLE_REQUEST,
+      createBlob: types.CREATE_BLOB,
+    }),
+
+    async submit () {
+      if (!this.isFormValid()) return
+
+      this.isDocumentUploading = true
+      this.disableForm()
+
+      try {
+        const documentText = await FileUtil.getText(this.form.document.file)
+        const documentHash = await CryptoUtil.hashMessage(documentText)
+
+        const accountId = await this.deriveAccountId(documentText)
+        await this.createAccount(accountId)
+
+        const documentId = await this.uploadDocument(
+          this.form.document,
+          this.wallet.accountId
+        )
+
+        const blobId = await this.createBlob({
+          id: documentId,
+          hash: documentHash,
+          description: this.form.description,
+        })
+
+        await this.createChangeRoleRequest(blobId)
+
+        Bus.success('document-upload-form.uploaded-msg')
+        this.$emit(EVENTS.submit)
+      } catch (e) {
+        if (e instanceof errors.BadRequestError) {
+          ErrorHandler.process(e, 'document-upload-form.document-exists-error')
+        } else {
+          ErrorHandler.process(e)
+        }
+      }
+
+      this.hideConfirmation()
+      this.isDocumentUploading = false
+      this.enableForm()
+    },
+
+    async deriveAccountId (documentText) {
+      const hashBuffer = await CryptoUtil.sha256(documentText)
```

Comment:
i don't like that we're doing it twice - when hashing the document and when calculating account id

Reply:
Updated, hashed the document only once, but used hash buffer instead of hex hash while hashing the document